### PR TITLE
GraphScope 2/3: (graph-learn on vineyard) use vineyard fragment as the underlying graph.

### PR DESCRIPTION
**Pull request by sighingnow**

### CODE REVIEW

File: graphlearn/src/service/dist/grpc_channel.cc [hunk 1]
```diff
@@ -106,7 +113,13 @@ Status GrpcChannel::CallDagValues(const DagValuesRequestPb* req,
 }
 
 Status GrpcChannel::CallStop(const StopRequestPb* req, StatusResponsePb* res) {
-  if (broken_) {
+  // TODO(tao): do we need such a check ?
+  //
+  // if (stopped_.load()) {
+  //   return Status::OK();
+  // }
+  stopped_.store(true);
```
Comment (Seventeen17) [inline graphlearn/src/service/dist/grpc_channel.cc:121]: The `stopped_` flag is to record the channel stopped state that I can see, so it seems that we should move the setting of `stopped_` after `HandleStop`

File: graphlearn/src/core/graph/storage/types.h [hunk 2]
```diff
@@ -32,39 +32,160 @@ typedef std::vector<IndexType> IndexList;
 typedef std::vector<std::vector<IdType>> IdMatrix;
 typedef std::unordered_map<IdType, IndexType> MAP;
 
+template <class T>
+class MultiArray {
+public:
+  MultiArray(const std::vector<const T*>& values, const std::vector<int32_t> sizes,
```
Comment (baoleai) [inline graphlearn/src/core/graph/storage/types.h:38]: What is the purpose of using MultiArray and RangeArray?

File: graphlearn/src/core/operator/sampler/alias_method.h [hunk 3]
```diff
@@ -52,6 +52,24 @@ class AliasMethodFactory {
...
+      auto am = new AliasMethod(&tmp_w);
+      map_[type] = am;
+      return am;
+    } else {
+      return it->second;
+    }
+  }
   
   template<class T>
   AliasMethod* LookupOrCreate(const std::string& type,
```
Comment (baoleai) [inline graphlearn/src/core/operator/sampler/alias_method.h:75]: Duplicate with line 57 `LookupOrCreate` ?

File: graphlearn/src/core/operator/sampler/condition_table.h [hunk 4]
```diff
@@ -122,6 +122,20 @@ class ConditionTableFactory {
     }
   }
 
+  template<class T>
+  ConditionTable* LookupOrCreate(
```
Comment (baoleai) [inline graphlearn/src/core/operator/sampler/condition_table.h:126]: Multiple duplicate interfaces, we only need to keep 2 according to the previous interface？

File: graphlearn/python/config.py [hunk 5]
```diff
@@ -106,4 +106,10 @@ def set_neg_sampler_retry_times(times):
   pywrap.set_neg_sampler_retry_times(times)
 
 def set_field_delimiter(delimiter="\t"):
-  pywrap.set_field_delimiter(delimiter)
\ No newline at end of file
+  pywrap.set_field_delimiter(delimiter)
+
+def set_vineyard_graph_id(graph_id):
```
Comment (baoleai) [inline graphlearn/python/config.py:111]: Better provide an example of the python side of vineyard?

File: graphlearn/src/include/data_source.h [hunk 6]
```diff
@@ -83,6 +83,9 @@ struct NodeSource {
   IndexOption   option;
   bool          local_shared;
 
+  std::string view_type;
```
Comment (baoleai) [inline graphlearn/src/include/data_source.h:86]: Are these two args only used in vineyard ? If so, why is it added to each graph creation function?

Comment (sighingnow) [reply graphlearn/src/include/data_source.h:86]: Yes, those two argument are only used by the vineyard implementation. But the dispatch happens inside the macro `CREATE(Type)` (see https://github.com/alibaba/graph-learn/blob/master/graphlearn/src/core/graph/storage_creator.cc#L21).

The macro was updated as follows when integrating with vineyard in 2020,

https://github.com/alibaba/graph-learn/blob/641b305d74ed93f82b692c76039999ec6a32e5e6/graphlearn/core/graph/storage_creator.cc#L21-L40

Thus we need to pass these two arguments util reaching the macro, even they are not used by other storage implementation.

File: graphlearn/python/config.py (hunk 5 above)
Comment (sighingnow) [reply graphlearn/python/config.py:111]: The python changes will be pushed in another PR (see overall plan in https://github.com/alibaba/graph-learn/pull/208#issue-1310757720).

File: graphlearn/src/service/dist/grpc_channel.cc (hunk 1 above)
Comment (sighingnow) [reply graphlearn/src/service/dist/grpc_channel.cc]: Resolved.

File: graphlearn/src/core/graph/graph.h [hunk 7]
```diff
@@ -46,8 +46,12 @@ class Graph {
 #undef DECLARE_METHOD
 };
 
-Graph* CreateLocalGraph();
-Graph* CreateRemoteGraph();
+Graph* CreateLocalGraph(const std::string& type,
+                        std::string const &view_type,
+                        std::string const &use_attrs);
```
[the remark this answers is not in the document]
Comment (sighingnow) [reply graphlearn/src/core/graph/graph.h]: Fixed.

File: graphlearn/src/core/operator/sampler/condition_table.h (hunk 4 above)
Comment (sighingnow) [reply graphlearn/src/core/operator/sampler/condition_table.h:126]: It is not duplicated. It is an overload for `io::Array<T>`. The reason for such an `io::Array` is that graphlearn internally use `std::vector<T>` for ids (`std::vector<int64_t>`) and datas (weights, `std::vector<float>`). However in vineyard those arrays are not `std::vector<T>` but a consecutive arrays on raw memory.

Thus we introduce the `io::Array` type for a view representation of the underlying raw memory. And we added overloads in `ConditionTable` and `AliasMethod`.

File: graphlearn/src/core/operator/sampler/alias_method.h (hunk 3 above)
Comment (sighingnow) [reply graphlearn/src/core/operator/sampler/alias_method.h:75]: It is an overloading, not a duplication. See my explaination for `condition_table.h`.

File: graphlearn/src/core/graph/storage/types.h (hunk 2 above)
Comment (sighingnow) [reply graphlearn/src/core/graph/storage/types.h]: The `MultiArray` is a view that treats multiple chunked array as a consective array without extra copying.

The `RangeArray` is a view that treats a range as a array without actually generating such an array to save memory.

File: graphlearn/CMakeLists.txt [hunk 8]
```diff
@@ -247,6 +255,17 @@ else ()
   set (GL_KNN_FILES)
 endif ()
 
+## contrib vineyard
+if (WITH_VINEYARD)
+  message(INFO "Graph-learn is built with vineyard support.")
+  find_package (vineyard)
```
Comment (baoleai) [inline graphlearn/CMakeLists.txt:261]: Should we also provide vineyard installation here or in `install_dependencies.sh`?
And we should add the installation process with vineyard  support to the documentation.

File: graphlearn/src/core/operator/sampler/condition_table.h (hunk 4 above)
Comment (baoleai) [reply graphlearn/src/core/operator/sampler/condition_table.h:126]: It appears that the current code does not use these classes.

File: graphlearn/src/core/graph/storage/types.h (hunk 2 above)
Comment (baoleai) [reply graphlearn/src/core/graph/storage/types.h]: It appears that the current code does not use these classes.

**Review by Seventeen17 — APPROVED** [review]
LGTM

File: graphlearn/CMakeLists.txt (hunk 8 above)
Comment (sighingnow) [reply graphlearn/CMakeLists.txt:261]: Vineyard will be available with `brew install` and `apt-get install` later.

I would update the installation docs then.

**Review by baoleai — APPROVED** [review]

**Pull request merged**